### PR TITLE
API Update - Allow user to configure a feed to fetch data from another Contao Portfolio

### PR DESCRIPTION
**Pull request by LupusVII**
So a lot of tiny adjustements for templates and data formating

### CODE REVIEW

File: src/Controller/ApiController.php [hunk 1]
```diff
@@ -100,53 +108,93 @@ public function viewPortfolioList(Request $request, int $page, int $limit, array
...
 
-        $cats = $request->query->all('cats');
+        $pid = $request->query->all('pid');
+        $lang = $request->query->get('lang') ?: $GLOBALS['TL_LANGUAGE'];
 
         $offset = ($page - 1) * $limit;
-        if (!is_iterable($cats)) {
-            return new JsonResponse('{"error":"Give at least on category : ?cats[]=1&cats[]=2"}', Response::HTTP_NOT_ACCEPTABLE, [], true);
+        if (!is_iterable($pid)) {
+            return new JsonResponse('{"error":"Give at least on category : ?pid[]=1&pid[]=2"}', Response::HTTP_NOT_ACCEPTABLE, [], true);
```
Comment (ChipsVII) [inline src/Controller/ApiController.php:116]: `at least one`

File: src/Module/ModulePortfolios.php [hunk 2]
```diff
@@ -208,12 +235,156 @@ protected function parsePortfolio(Portfolio $objItem, bool $blnAddArchive = fals
...
+        curl_close($ch);
+        $data = json_decode($request, true);
+
+        // We need to format a Collection of Portfolio
+        $items = [];
+        foreach ($data as $item) {
+            $objModel = new Portfolio();
+            $objModel->pid = $feed->id;
+
+            foreach ($item as $c => $v) {
```
Comment (ChipsVII) [inline src/Module/ModulePortfolios.php:272]: why not 
```php
unset($item['category'],$item['pid']); // neither category nor pid should be set/overriden
$objModel->setRow($item);
```

File: src/Module/ModulePortfolios.php [hunk 3]
```diff
@@ -208,12 +235,156 @@ protected function parsePortfolio(Portfolio $objItem, bool $blnAddArchive = fals
...
+        $url = $feed->readFromRemoteUrl . '/api/portfolio/count?' . $params;
+
+        curl_setopt($ch, CURLOPT_URL, $url);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+
+        $request = curl_exec($ch);
+        curl_close($ch);
+        $data = json_decode($request, true);
+
+        return $data['items'];
```
Comment (ChipsVII) [inline src/Module/ModulePortfolios.php:315]: `return (int) $data['items'];`

File: src/Module/ModulePortfolios.php [hunk 4]
```diff
@@ -208,12 +235,156 @@ protected function parsePortfolio(Portfolio $objItem, bool $blnAddArchive = fals
...
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+
+        $request = curl_exec($ch);
+        curl_close($ch);
+        $data = json_decode($request, true);
+
+        $objModel = new Portfolio();
+        $objModel->pid = $feed->id;
+
+        foreach ($data as $c => $v) {
```
Comment (ChipsVII) [inline src/Module/ModulePortfolios.php:345]: same as above, why not using `$objModel->setRow($data);` ?

File: src/Module/ModulePortfolios.php [hunk 5]
```diff
@@ -208,12 +235,156 @@ protected function parsePortfolio(Portfolio $objItem, bool $blnAddArchive = fals
...
+
+                default:
+                    $objModel->{$c} = $v;
+            }
+        }
+
+        return $objModel;
+    }
+
+    protected function formatConfigForRemote($config, $feed): string
```
Comment (ChipsVII) [inline src/Module/ModulePortfolios.php:359]: Parameters typing if possible ?

File: src/Resources/contao/dca/tl_wem_portfolio_feed.php [hunk 6]
```diff
@@ -121,6 +128,38 @@
...
+            'filter' => true,
+            'flag' => 1,
+            'inputType' => 'checkbox',
+            'eval' => ['doNotCopy' => true, 'submitOnChange' => true],
+            'sql' => "char(1) NOT NULL default ''",
+        ],
+        'readFromRemoteUrl' => [
+            'exclude' => true,
+            'inputType' => 'text',
+            'eval' => ['rgxp'=>'url', 'decodeEntities'=>true, 'maxlength'=>2048, 'dcaPicker'=>true, 'tl_class'=>'w50'],
```
Comment (ChipsVII) [inline src/Resources/contao/dca/tl_wem_portfolio_feed.php:142]: What does `'dcaPicker'=>true` do in that case ? Contao's doc is quite shy on details.

File: src/Controller/ApiController.php [hunk 7]
```diff
@@ -100,53 +108,93 @@ public function viewPortfolioList(Request $request, int $page, int $limit, array
...
-                        $return['mainPicture'] = [];
-
-                        if ('1' === $arrayItem['addImage']) {
-                            $imageP = FilesModel::findByUuid($arrayItem['singleSRC']);
-                            $uuidP = Uuid::fromBinary($imageP->uuid);
-                            $return['image_principal'][$uuidP->__toString()]['path'] = $imageP->path;
-                            $return['image_principal'][$uuidP->__toString()]['tstamp'] = $imageP->tstamp;
-                            $return['image_principal'][$uuidP->__toString()]['hash'] = $imageP->hash;
-                            $return['image_principal'][$uuidP->__toString()]['lastModified'] = $imageP->lastModified;
+                        foreach($arrayItem as $c => $v) {
```
[the remark this answers is not in the document]
Comment (LupusVII) [reply src/Controller/ApiController.php]: Fixed in 5ff0bedc0429915c064e1a13e56e562c45b50ff2

File: src/Controller/ApiController.php (hunk 1 above)
Comment (LupusVII) [reply src/Controller/ApiController.php]: Fixed in 5ff0bedc0429915c064e1a13e56e562c45b50ff2

File: src/Module/ModulePortfolios.php (hunk 4 above)
Comment (LupusVII) [reply src/Module/ModulePortfolios.php]: Fixed in 22fd2c0754b679ba12fdae14a9387e7c7a328871

File: src/Module/ModulePortfolios.php (hunk 3 above)
Comment (LupusVII) [reply src/Module/ModulePortfolios.php]: Fixed in afcdbb2b0f84ea2ea466b6a7b8b36a28f1965a27

File: src/Module/ModulePortfolios.php (hunk 5 above)
Comment (LupusVII) [reply src/Module/ModulePortfolios.php]: Fixed in fe448c71f043da604ab1280d6a3036f01def9948

File: src/Resources/contao/dca/tl_wem_portfolio_feed.php (hunk 6 above)
Comment (LupusVII) [reply src/Resources/contao/dca/tl_wem_portfolio_feed.php]: Wrong copy/paste, fixed in 4c83f666ea75b8a2b2f54949d18540e9dd8d22e3

**Pull request merged**